### PR TITLE
WT-9811 Fast-truncate page for the architecture guide

### PR DESCRIPTION
**Pull request by sauclovian-wt**
Add a fast-truncate page to the architecture guide.

Improve s_docs so we don't need non-words appearing in certain filenames and symbols (e.g. "vrfy") to be spelling words. This is done with ad hoc seddery because there are not many cases and not that many new ones likely to appear.

Add some spelling words.

Drop most of the old brief description in arch-btree and refer to the new page instead.

Also update the discussion of the transactionality of truncate operations. I'm not changing what the user-facing docs say without discussion, but we can afford to document the current understanding here instead of having it buried in various tickets and github comments.

(The known bug referred to is WT-4158. I didn't put the ticket number in the docs because that doesn't seem like the right thing, but can certainly add it if desired.)

### CODE REVIEW

**Comment by sauclovian-wt** [comment]
Note that there are probably a thousand typos in this because proofreading one's own text is never that effective. If marking them up with the review interface gets annoying, feel free to just push fixes for any that are obvious and don't require questions or discussion, which will probably be most of them.

Also, I'm happy to rearrange the table of manifestations at the bottom if anyone thinks up a better way to present that information. (I think it's important to have it, even though it'll doubtless go out of date fairly rapidly.)

**Comment by sauclovian-wt** [comment]
Oh, also note that I wrote this with new-sentence-new-line (always begin every sentence on a new line) because this makes diffs much more readable. This is not (yet) general WT practice, but I'd encourage adopting it for docs; we've found in NetBSD that it helps a lot.

**Comment by etienneptl** [comment]
Thanks for those changes @sauclovian-wt, I am adding myself as a reviewer as I am interested. I will review asap.

**Comment by agorrod** [comment]
The testing failed with:

```
[2022/08/28 11:22:19.336] ####################### MESSAGE ############################
[2022/08/28 11:22:19.336] s_all run of: "./s_docs" resulted in:
[2022/08/28 11:22:19.337]     /data/mci/b9e5bb04f5dcaf410adc5e2ec8981c9d/wiredtiger/src/docs/arch-fast-truncate.dox:1069: error: Reached end of file while still inside a (nested) comment. Nesting level 1 (probable line reference: 1) (warning treated as error, aborting now)
[2022/08/28 11:22:19.343] #######################
```

It isn't trivial for me to checkout your fork and update it - so I'll leave it for you to fix if that's OK?

**Comment by etienneptl** [comment]
> The testing failed with:
> 
> ```
> [2022/08/28 11:22:19.336] ####################### MESSAGE ############################
> [2022/08/28 11:22:19.336] s_all run of: "./s_docs" resulted in:
> [2022/08/28 11:22:19.337]     /data/mci/b9e5bb04f5dcaf410adc5e2ec8981c9d/wiredtiger/src/docs/arch-fast-truncate.dox:1069: error: Reached end of file while still inside a (nested) comment. Nesting level 1 (probable line reference: 1) (warning treated as error, aborting now)
> [2022/08/28 11:22:19.343] #######################
> ```
> 
> It isn't trivial for me to checkout your fork and update it - so I'll leave it for you to fix if that's OK?

Indeed, it misses a curly bracket at the end of [this line](https://github.com/wiredtiger/wiredtiger/pull/8237/files#diff-ab5bb352801d56c239dbe3e80b695f24beec528c86910c48310107a44655da84R641):

```
@row{btree_inline.h, Hook, in \c __wt_ref_addr_copy,
A hook to return the page delete information unpacked from the address
```

File: src/docs/arch-btree.dox [hunk 1]
```diff
@@ -54,34 +54,41 @@ the focus here will be on truncation for B-Tree data files (\c file: uri).
...
-that both commit, if the system crashes and recovery runs, the resulting state of the
-database may be different to what was in the cache had the crash not happened.
+Historically we have stated that truncation is not transactional: if a range truncate is in
+progress and another transaction happens to insert a key into the same range, the behavior is
+not well defined.
+This was a hedge for a known bug; that bug is now understood, and it is limited to cases in
+logged trees where the truncation passes \c NULL instead of a cursor for the start and/or end
+point and that point changes concurrently; it is possible in these circumstances for log
+replay to find and use a different position for the start or end of the tree than occurred in
+the original transaction.
```
Comment (michaelcahill) [inline src/docs/arch-btree.dox:78]: I agree that the problem applies to logged trees with recovery, but unless things have changed, I think it is broader than described above.

I think the problem is that you can truncate [A, C] in T1 and insert B in a concurrent T2. Depending on how T1 and T2 interleave, you could have the following outcomes:

1. T2 inserts B, T1 visits page containing B and gets a write conflict;
2. T1 goes past B, T2 inserts B, T1 commits, T2 commits — no conflict during execution, B successfully inserted
3. T1 goes past B, T2 inserts B, T2 commits, T1 commits — no conflict during execution, B successfully inserted, but it will be silently removed during recovery when T2 is applied after T1.

Since applications can’t readily distinguish between cases 2 and 3 (and the state of 3 after recovery is different to the original execution), the documentation tries to say “don’t do that”.

I’m not sure what’s special about the case where A=NULL (or presumably C=NULL?) — I always thought the issue was that recovery could produce a different result than was seen originally.

**Comment by sauclovian-wt** [comment]
On the failure, I'm very confused. It definitely built several times after git says that line was last changed in my working tree, and I'd swear it built when I pushed it, but it doesn't build now...

**Comment by sauclovian-wt** [comment]
Anyway, fixed.

File: src/docs/arch-btree.dox (hunk 1 above)
Comment (sauclovian-wt) [reply src/docs/arch-btree.dox:78]: If it's an insert and not an update, (1) can't happen (the new value isn't visible to the truncate if they're running concurrently) but the point about recovery order is a good one that I'd previously missed. That is: yes, because truncate only logs the range, write phantoms might disappear after recovery. I'll add that to the notes in WT-4158.

What's special about NULL is that we log NULL instead of resolving the endpoint and logging that; recovery then figures out the end of the table for itself and it's possible for it to get a different answer. I discovered this behavior when debugging WT-4158 a while back.

(And I'll update the text here)

File: src/docs/arch-fast-truncate.dox [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+page is thenceforth a completely normal in-memory page.
+
+Note that instantiated pages are not automatically marked dirty.
+(See the notes on this point in the @ref ft_instantiation section.)
+
+Also note that it is not possible to fast-truncate an instantiated page.
+Only on-disk pages can be fast-truncated.
+See @ref ft_truncate.
+
+@subsection ft_instantiated_instantiated Instantiated flag and \c page_del
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:118]: I generated the doc and it seems that `\c` does not get processed for headers. The same happens with the next header `inst_updates`

File: src/docs/arch-fast-truncate.dox [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+
+The instantiation code iterates the page and adds a tombstone for every item.
+(To be precise, it adds a tombstone for every item that isn't already deleted; items that
+have a stop time do not need to be deleted again.)  The row-store version of this iterates
+the entries on the page and directly adds a tombstone to each update list.
+The column-store version uses a cursor and calls into \c col_modify.c to do it; this is
+because the data structures are considerably more complicated and updating them directly
+would require a lot of cut-and-paste code.
+
+The tombstones are tagged with WT_UPDATE_RESTORED_FAST_TRUNCATE.
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:181]: We probably want a `\c` before `WT_UPDATE_RESTORED_FAST_TRUNCATE`

File: src/docs/arch-fast-truncate.dox [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+its items explicitly deleted with tombstone updates.
+<i>Semantically, this is an identity transformation.</i>
+
+This occurs under two sets of circumstances: first, if a thread that cannot yet see the
+deletion tries to read from the deleted page; and second, if a search lands on the deleted
+page's portion of the namespace.
+It can happen either before or after the transaction that deleted the page is resolved.
+(But not at the same time; locking the ref prevents that.)
+
+(Note that for searches that are positioning to do updates, the instantiation is
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:160]: This whole paragraph seems too big and useful to be between bracket, we can probably remove them :)

File: src/docs/arch-fast-truncate.dox [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+unavoidable; however, for searches that are only reading, it would be better to return \c
+WT_NOTFOUND without pointlessly instantiating the page.
+This is not currently implemented and as of this writing it is not entirely clear how
+involved or feasible such changes would be.
+Also note that this only applies to explicit searches and searches that are part of other
+cursor operations; the cursor next and previous operations do skip over deleted pages.)
+
+In either case the fact that the page is deleted is detected when we read the page.
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:168]: I am not sure about this one but I would reword it:
```suggestion
In either case, the deleted page is detected when we read it.
```

File: src/docs/arch-fast-truncate.dox [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+The instantiation code iterates the page and adds a tombstone for every item.
+(To be precise, it adds a tombstone for every item that isn't already deleted; items that
+have a stop time do not need to be deleted again.)  The row-store version of this iterates
+the entries on the page and directly adds a tombstone to each update list.
+The column-store version uses a cursor and calls into \c col_modify.c to do it; this is
+because the data structures are considerably more complicated and updating them directly
+would require a lot of cut-and-paste code.
+
+The tombstones are tagged with WT_UPDATE_RESTORED_FAST_TRUNCATE.
+This is used by prepare to avoid trying to coalesce these updates with others to the same
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:182]: By `prepare`, do you mean `prepare_transaction`? If that's the case, can we be more explicit and add a `\c` as well?

File: src/docs/arch-btree.dox (hunk 1 above)
Comment (sauclovian-wt) [reply src/docs/arch-btree.dox]: I have improved what's posted in WT-4158. The text here is fixed in [02cbfa9](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/02cbfa981bf3b19022a5d8fcd84491bb833b3d2f).

File: src/docs/arch-fast-truncate.dox (hunk 2 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Boo on doxygen :-(  Apparently you can't use the raw tt tag either. Fixed in [f3a555c](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/f3a555cfd7ddc3655258d577632287a485c5676a)

File: src/docs/arch-fast-truncate.dox (hunk 3 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Yup, good catch, fixed in [f3a555c](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/f3a555cfd7ddc3655258d577632287a485c5676a)

File: src/docs/arch-fast-truncate.dox (hunk 4 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Removed them in [f3a555c](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/f3a555cfd7ddc3655258d577632287a485c5676a).

There is no ticket that I know of. It is on my private "would be nice sometime" list along with other expensive and impractical things. If you think it's worth filing a ticket on, I can do that...

File: src/docs/arch-fast-truncate.dox (hunk 5 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: I'm going to push back slightly on that; we're detecting the page state, not the page. However, I agree that the existing sentence is awful! I'll come up with something else... how about "In any case, we notice the page is deleted when we read it."? This is in [f3a555c](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/f3a555cfd7ddc3655258d577632287a485c5676a), can obviously be changed further if desired.

File: src/docs/arch-fast-truncate.dox (hunk 6 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: I meant "prepared transaction handling" but it's actually in `__wt_txn_prepare` and I might as well be explicit. Fixed in [f3a555c](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/f3a555cfd7ddc3655258d577632287a485c5676a)

File: src/docs/arch-btree.dox [hunk 7]
```diff
@@ -54,34 +54,49 @@ the focus here will be on truncation for B-Tree data files (\c file: uri).
...
+behavior one must recognize that truncate is a read-write operation: it scans the tree from the
+start key to the end key and deletes every value it finds. Concurrent updates where the same
+scan and update done manually would report a conflict cause truncate to report a conflict.
+Concurrent updates that this scan would miss (for example, insertion of new rows in a
+row-store tree) will be skipped over by the truncate and not cause a conflict. (This is a
+form of <i>phantom</i> behavior and is a result of the isolation model, not a transaction
+bug.)
+The second logging bug is that these phantom writes, which should be left alone, are removed by log
+recovery if the transaction that created them commits before the truncate.
+The logging for truncate saves the endpoints of the truncate range,, not the individual
```
Comment (etienneptl) [inline src/docs/arch-btree.dox:90]: Extra `,` here on this line:
```suggestion
The logging for truncate saves the endpoints of the truncate range, not the individual
```

Comment (sauclovian-wt) [reply src/docs/arch-btree.dox]: Fixed in [67208a8](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/67208a8ebbd0b18b300fa31f4e3b1063779e1d4b), thanks.

File: src/docs/arch-fast-truncate.dox [hunk 8]
```diff
@@ -0,0 +1,847 @@
...
+Truncations that are globally visible do not result in a cell in the parent page at all.)
+
+These fields are, however, only present if the page header includes the \c
+WT_PAGE_FT_UPDATE flag, whose value is 0x20.
+Proper support for timestamped fast-truncate only appeared in WT 11.0; earlier versions
+neither write these fields nor expect to find them.
+The explicit header flag is required to make compatibility guarantees function as needed.
+(MongoDB 6.x knows how to read pages with \c WT_PAGE_FT_UPDATE set, but does not write
+them.
+This is controlled in the WT library by the build config.)
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:336]: @sauclovian-wt, can you elaborate on this? I am not sure to understand. What building parameter(s) need(s) to be changed?

File: src/docs/arch-fast-truncate.dox [hunk 9]
```diff
@@ -0,0 +1,847 @@
...
+Fast-truncate then happens inside the tree-walk code (in \c bt_walk.c) which calls \c
+__wt_delete_page on leaf pages it visits.
+
+This function, in \c bt_delete.c, checks the page for eligibility.
+First, if the page is unmodified and in memory we attempt to evict it.
+Then we check if the page is on disk, that is, the state is \c WT_REF_DISK, and if so lock
+the ref.
+(And then check if it is <i>still</i> on disk.)
+
+Then the following categories of pages are ineligible for fast-truncate:
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:356]: Is it possible to have bullet points or a simple list for the different reasons listed in this paragraph?

File: src/docs/arch-fast-truncate.dox [hunk 10]
```diff
@@ -0,0 +1,847 @@
...
+ft_vlcs.
+
+Finally, new trees are created with a single empty deleted leaf page, because creating the
+tree with no leaves at all causes problems.
+
+@section ft_page_skip Skipping deleted pages
+
+There are two separate skip functions for skipping deleted pages during tree walks.
+The basic skip function is \c __wt_delete_page_skip in \c bt_delete.c; it is always called
+in tree walks when \c WT_READ_SKIP_DELETED is set, which is all tree walks except RTS.
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:422]: Can we point to the RTS page `arch-rts`?

File: src/docs/arch-fast-truncate.dox [hunk 11]
```diff
@@ -0,0 +1,847 @@
...
+The other skip function is \c __wt_btcur_skip_page in \c btree_inline.h.
+This is used by cursor next and previous to skip over deleted pages in those traversals
+specifically, and in addition to checking for \c WT_REF_DELETED with visible \c page_del,
+it also inspects the address cell.
+If the page is in memory and unmodified and the address cell contains page-delete
+information, it checks for the visibility of that deletion.
+Failing that, it checks the time aggregate in case all the values on the page are no
+longer visible, e.g. because the page was slow-truncated and reconciled.
+
+(For an unmodified in-memory page the deletion information in the address should be the
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:439]: I think we can drop the brackets here too for this whole paragraph.

File: src/docs/arch-fast-truncate.dox [hunk 12]
```diff
@@ -0,0 +1,847 @@
...
+principle it's possible to drop the ref's \c page_del immediately upon instantiation in a
+read-only tree, since it is kept for internal page reconciliation and pages in read-only
+trees are never reconciled.
+Checking the unpacked information here avoids interfering with that option.)
+
+Note that \c __wt_btcur_skip_page is passed to the tree-walk code as a custom skip
+function, which means that when it's used <i>both</i> it and \c __wt_delete_page_skip are
+checked.
+This is not optimal and probably ought to be tidied.
+Furthermore, it is unclear whether these functions really need to be different; that is,
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:454]: Should we create a ticket for this investigation ?

File: src/docs/arch-fast-truncate.dox [hunk 13]
```diff
@@ -0,0 +1,847 @@
...
+@row{reconcile_inline.h, Hook, in \c __wt_rec_cell_build_addr,
+A hook to choose \c WT_CELL_ADDR_DEL cells when needed and propagate any passed-in page
+deletion information to the packing code.}
+
+@row{timestamp_inline.h, Macro, \c WT_TIME_AGGREGATE_UPDATE_PAGE_DEL,
+Akin to \c WT_TIME_AGGREGATE_UPDATE but for a \c WT_PAGE_DELETED; used in internal page
+reconciliation.}
+
+@row{txn_inline.h, Function, \c __wt_txn_op_delete_apply_prepare_state,
+The code for updating \c WT_PAGE_DELETED at prepare time.  updates to \c WT_PAGE_DELETED.}
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:664]: Is this correct?

File: src/docs/arch-fast-truncate.dox [hunk 14]
```diff
@@ -0,0 +1,847 @@
...
+@row{txn_inline.h, Function, \c __wt_txn_op_commit_apply_timestamps,
+Part of the code for updating \c WT_PAGE_DELETED at commit time.
+(The rest is in \c __wt_txn_commit itself.
+Note that the rollback-time update is in bt_delete.c.)}
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:668]: We can add "\c" before `bt_delete.c`

File: src/docs/arch-fast-truncate.dox [hunk 15]
```diff
@@ -0,0 +1,847 @@
...
+@row{txn_inline.h, Hook, in \c __wt_txn_op_set_timestamp,
+Call \c __wt_txn_op_delete_apply_prepare_state and
+\c __wt_txn_op_commit_apply_timestamps.}
+@row{txn_inline.h, Function, \c __wt_txn_modify_page_delete,
+This records a fast-truncate in the current transaction.}
+
+@row{cell_inline.h, Function, \c __cell_page_del_window_cleanup,
+Akin to \c __cell_kv_window_cleanup except for \c WT_PAGE_DELETED.}
+@row{cell_inline.h, Function, \c __cell_redo_page_del_cleanup,
+Function that redoes does the timestamp cleanup for a \c WT_PAGE_DELETED structure; used
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:678]: "redoes" or "does"? :)

File: src/docs/arch-fast-truncate.dox [hunk 16]
```diff
@@ -0,0 +1,847 @@
...
+@row{bt_curnext.c, Hook, in \c __wt_btcur_next,
+Accept a flag argument that sets \c WT_READ_TRUNCATE.}
+
+@row{bt_curprev.c, Hook, in \c __wt_btcur_prev,
+Accept a flag argument that sets \c WT_READ_TRUNCATE.}
+
+@row{bt_cursor.c:, Function, \c __wt_btcur_range_truncate (and others),
+The cursor-level truncate code lives here.}
+
+@row{bt_debug.c, Hook, in \c __debug_cell_int,
```
Comment (etienneptl) [inline src/docs/arch-fast-truncate.dox:696]: Should we create a ticket for those two?

File: src/docs/arch-fast-truncate.dox (hunk 8 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Currently it's controlled by WT_STANDALONE_BUILD, indirectly by a flag in __wt_process. I didn't want to be too explicit because it might change, but I'll fix it to at least name the flag. (In [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466))

File: src/docs/arch-fast-truncate.dox (hunk 9 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox:357]: It is supposed to be! The doxygen manual says that's how to make a bullet list. Apparently it's just lying and you need to use - and not *. Fixed in [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466). Also, I fixed up the wording of the overflow items entry.

File: src/docs/arch-fast-truncate.dox (hunk 10 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Good thought. Done in [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466)

File: src/docs/arch-fast-truncate.dox (hunk 11 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Sure. In [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466)

File: src/docs/arch-fast-truncate.dox (hunk 12 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox:457]: I guess so... WT-9859

File: src/docs/arch-fast-truncate.dox (hunk 13 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Apparently not :-) The second half appears to be some sort of leftover. Removed in [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466)

File: src/docs/arch-fast-truncate.dox (hunk 14 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Fixed in [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466)

File: src/docs/arch-fast-truncate.dox (hunk 15 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox]: Fixed :-)

File: src/docs/arch-fast-truncate.dox (hunk 16 above)
Comment (sauclovian-wt) [reply src/docs/arch-fast-truncate.dox:700]: I guess so. WT-9860

File: src/docs/arch-fast-truncate.dox (hunk 9 above)
Comment (etienneptl) [reply src/docs/arch-fast-truncate.dox:357]: The bullet point list is working now! And thanks for the rewording.

**Review by etienneptl — APPROVED** [review]
Excellent documentation, I learned a lot, thanks for doing this!

**Comment by sauclovian-wt** [comment]
YW :-)

**Review by agorrod — APPROVED** [review]
Thanks @sauclovian-wt - I particularly appreciate the reference section.

I think this is good to go - so I'll go ahead and merge.

**Review by michaelcahill — APPROVED** [review]
Thanks @sauclovian-wt, lgtm.

**Comment by sauclovian-wt** [comment]
YW all :-)
